buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -0,0 +1,1428 @@
+                        buffer.release();
+                    }
+                }
+            }
+            return result;
+        }
+
+        private ChunkByteBuf newChunkAllocation(int promptingSize) {
+            int size = Math.max(promptingSize * BUFS_PER_CHUNK, preferredChunkSize());
+            ChunkAllocator chunkAllocator = parent.chunkAllocator;
+            usedMemory.getAndAdd(size);
+            ChunkByteBuf chunk = new ChunkByteBuf((AbstractByteBuf) chunkAllocator.allocate(size, size), this);
+            chunk.writerIndex(size);
+            return chunk;
+        }
+
+        boolean trySetNextInLine(ChunkByteBuf buffer) {
+            return NEXT_IN_LINE.compareAndSet(this, null, buffer);
+        }
+
+        void close() {
+            ByteBuf curr = current;
+            if (curr != null) {
+                current = null;
+                curr.release();
+            }
+            curr = NEXT_IN_LINE.getAndSet(this, null);
+            if (curr != null) {
+                curr.release();
+            }
+        }
+    }
+
+    private static final class ChunkByteBuf extends AbstractReferenceCountedByteBuf {
+        private final AbstractByteBuf delegate;
+        private final Magazine magazine;
+
+        ChunkByteBuf(AbstractByteBuf delegate, Magazine magazine) {
+            super(delegate.maxCapacity());
+            this.delegate = delegate;
+            this.magazine = magazine;
+        }
+
+        @Override
+        protected void deallocate() {
+            Magazine mag = magazine;
+            AdaptivePoolingAllocator parent = mag.parent;
+            int chunkSize = mag.preferredChunkSize();
+            int memSize = delegate.capacity();
+            if (parent.closed || memSize < chunkSize || memSize > chunkSize + (chunkSize >> 1)) {
+                // Drop the chunk if the parent allocator is closed, or if the chunk is smaller than the
+                // preferred chunk size, or over 50% larger than the preferred chunk size.
+                mag.usedMemory.getAndAdd(-capacity());
+                delegate.release();
+            } else {
+                resetRefCnt();
+                delegate.setIndex(0, 0);
+                setIndex(0, capacity());
+                if (!mag.trySetNextInLine(this)) {
+                    if (!parent.offerToQueue(this)) {
+                        // The central queue is full. Drop the memory with the original Drop instance.
+                        delegate.release();
+                    }
+                }
+            }
+        }
+
+        public AdaptiveByteBuf readInitInto(AdaptiveByteBuf buf, int size, int maxCapacity) {
+            int startIndex = readerIndex();
+            skipBytes(size);
+            buf.init(delegate, this, 0, 0, startIndex, size, maxCapacity);
+            return buf;
+        }
+
+        @Override
+        protected byte _getByte(int index) {
+            return delegate.getByte(index);
```

Comment:
hmm... shouldn't we call the `._getByte(...)` method here so we don't pay for the extra bound checks. Same is true for all the other methods.